File: components/bunch/tests/test_diagnostics_writer.cc
#define BOOST_TEST_MAIN
#include <boost/test/unit_test.hpp>
#include "components/bunch/diagnostics_writer.h"
#include "components/foundation/physical_constants.h"
#include "utils/boost_test_mpi_fixture.h"
BOOST_GLOBAL_FIXTURE(MPI_fixture)
;

const double tolerance = 1.0e-12;

const double mass = 100.0;
const double total_energy = 125.0;
const int total_num = 9;
const double real_num = 2.0e12;
const int turns = 17;
const double turn_length = 246.8;
const double partial_s = 123.4;

void
dummy_populate(Bunch &bunch)
{
    for (int part = 0; part < bunch.get_local_num(); ++part) {
        for (int i = 0; i < 6; i += 1) {
            bunch.get_local_particles()[part][i] = 10.0 * part + (1.0 + part
                    * part / 1000.0) * i;
        }
        bunch.get_local_particles()[part][Bunch::id] = part;
    }
}

struct Fixture
{
    Fixture() :
        reference_particle(constants::electron_charge, mass, total_energy),
                comm(MPI_COMM_WORLD), bunch(reference_particle, total_num,
                        real_num, comm)
    {
        BOOST_TEST_MESSAGE("setup fixture");
        dummy_populate(bunch);
        bunch.get_reference_particle().set_trajectory(turns, turn_length,
                partial_s);
    }
    ~Fixture()
    {
        BOOST_TEST_MESSAGE("teardown fixture");
    }

    Reference_particle reference_particle;
    Commxx comm;
    Bunch bunch;
};

BOOST_FIXTURE_TEST_CASE(construct, Fixture)
{
    Diagnostics_sptr diagnostics_sptr(new Diagnostics(bunch));
    Diagnostics_writer diagnostics_writer("test_writer_construct.h5",
            diagnostics_sptr);
}

BOOST_FIXTURE_TEST_CASE(construct_full2, Fixture)
{
    Diagnostics_full2_sptr diagnostics_sptr(new Diagnostics_full2(bunch));
    Diagnostics_writer diagnostics_writer("test_writer_construct_full2.h5",
            diagnostics_sptr);
}

BOOST_FIXTURE_TEST_CASE(get_diagnostics_sptr, Fixture)
{
    Diagnostics_sptr diagnostics_sptr(new Diagnostics(bunch));
    Diagnostics_writer diagnostics_writer(
            "test_writer_get_diagnostics_sptr.h5", diagnostics_sptr);
    Diagnostics_sptr retrieved_d_s = diagnostics_writer.get_diagnostics_sptr();
    BOOST_CHECK_EQUAL(diagnostics_sptr, retrieved_d_s);
}

BOOST_FIXTURE_TEST_CASE(write_, Fixture)
{
    Diagnostics_sptr diagnostics_sptr(new Diagnostics(bunch));
    Diagnostics_writer diagnostics_writer("test_writer_write.h5", diagnostics_sptr);
    diagnostics_writer.get_diagnostics_sptr()->update(bunch);
    diagnostics_writer.write();
}

BOOST_FIXTURE_TEST_CASE(update_and_write, Fixture)
{
    Diagnostics_sptr diagnostics_sptr(new Diagnostics(bunch));
    Diagnostics_writer diagnostics_writer("test_writer_update_and_write.h5", diagnostics_sptr);
    diagnostics_writer.update_and_write(bunch);
}

BOOST_FIXTURE_TEST_CASE(update_and_write_full2, Fixture)
{
    Diagnostics_full2_sptr diagnostics_sptr(new Diagnostics_full2(bunch));
    Diagnostics_writer diagnostics_writer("test_writer_update_and_write_full2.h5", diagnostics_sptr);
    diagnostics_writer.update_and_write(bunch);
}

File: components/bunch/tests/test_diagnostics_writer.py
import sys
sys.path.append('..')
sys.path.append('../../foundation')
sys.path.append('../../convertors')

from mpi4py import MPI
from pyfoundation import Reference_particle, Four_momentum
from pybunch import Bunch
from pybunch import Diagnostics, Diagnostics_full2
from pybunch import Diagnostics_writer, no_diagnostics
import pyconvertors
import numpy
from nose.tools import *

mass = 2.2;
total_energy = 3.0
total_num = 100
real_num = 1.0e12
proton_charge = 1
turns = 17;
turn_length = 246.8;
partial_s = 123.4;


reference_particle = Reference_particle(proton_charge, mass, total_energy)
reference_particle.set_trajectory(turns, turn_length, partial_s)
comm = MPI.COMM_WORLD
bunch = Bunch(reference_particle, total_num, real_num, comm)
particles = bunch.get_local_particles()
particles[:, 0:6] = numpy.random.lognormal(size=[bunch.get_total_num(), 6])

def test_construct():
    diagnostics = Diagnostics()
    diagnostics_writer = Diagnostics_writer("test_py_construct.h5", diagnostics)

def test_construct_full2():
    diagnostics = Diagnostics_full2()
    diagnostics_writer = Diagnostics_writer("test_py_construct_full2.h5", diagnostics)

def test_construct_dummy():
    dummy = Diagnostics_writer()

def test_is_dummy():
    assert(Diagnostics_writer().is_dummy())

def test_get_diagnostics():
    diagnostics = Diagnostics()
    diagnostics_writer = Diagnostics_writer("test_py_get_diagnostics.h5", diagnostics)
    retrieved_diagnostics = diagnostics_writer.get_diagnostics()
    assert(type(retrieved_diagnostics) == type(diagnostics))

def test_write():
    diagnostics = Diagnostics(bunch)
    diagnostics_writer = Diagnostics_writer("test_py_write.h5", diagnostics)
    diagnostics_writer.write()

def test_update_and_write():
    diagnostics = Diagnostics()
    diagnostics_writer = Diagnostics_writer("test_py_update_and_write.h5", diagnostics)
    diagnostics_writer.update_and_write(bunch)
    diagnostics_writer.update_and_write(bunch)

def test_update_and_write_full2():
    diagnostics = Diagnostics_full2()
    diagnostics_writer = Diagnostics_writer("test_py_update_and_write_full2.h5", diagnostics)
    diagnostics_writer.update_and_write(bunch)
    diagnostics_writer.update_and_write(bunch)

def test_no_diagnostics():
    assert(no_diagnostics().is_dummy())
